### PR TITLE
RAT Create Profile Screen

### PR DESCRIPTION
**Pull request by freshking**
## Description

Create RAT Profile Screen with all input fields

## Link to Jira

https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-6619

## Screenshots

IMG_2945

IMG_2946

### CODE REVIEW

File: src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift [hunk 1]
```diff
@@ -29,11 +34,75 @@ class CreateAntigenTestProfileViewController: UIViewController, FooterViewHandli
...
+		return 1
+	}
+	
+	override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+		return 2
+	}
+	
+	override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+		switch indexPath.row {
+		case 0:
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift:51]: can we better use an enum to make thing more readable

File: src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift [hunk 2]
```diff
@@ -49,15 +118,290 @@ class CreateAntigenTestProfileViewController: UIViewController, FooterViewHandli
...
+		return true
+	}
 
 	// MARK: - Private
 
 	private let viewModel: CreateAntigenTestProfileViewModel
 	private let didTapSave: () -> Void
 	private let dismiss: () -> Void
+	
+	private var dateFormatter: DateFormatter!
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift:143]: I think we don't need it. You can use `DateFormatter.localizedString(from: date, dateStyle: .medium, timeStyle: .medium)`  for the default output

File: src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift [hunk 3]
```diff
@@ -49,15 +118,290 @@ class CreateAntigenTestProfileViewController: UIViewController, FooterViewHandli
...
 	}
+	
+	// MARK: - Protocol UITextFieldDelegate
 
-	// MARK: - Public
-
-	// MARK: - Internal
+	func textFieldShouldReturn(_ textField: UITextField) -> Bool {
+		if textField.returnKeyType == .next {
+			if let cell = tableView.cellForRow(at: IndexPath(row: 1, section: 0)) as? InputCell, let currentIndex = cell.textFields.firstIndex(of: textField), (currentIndex + 1) < cell.textFields.count {
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift:126]: Would be nice to have every "let .. =" in a new line.

File: src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift [hunk 4]
```diff
@@ -49,15 +118,290 @@ class CreateAntigenTestProfileViewController: UIViewController, FooterViewHandli
...
 		dismiss()
 	}
+	
+	// MARK: - Protocol UITextFieldDelegate
 
-	// MARK: - Public
-
-	// MARK: - Internal
+	func textFieldShouldReturn(_ textField: UITextField) -> Bool {
+		if textField.returnKeyType == .next {
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift:125]: Maybe using guard would reduce the pyramid.

File: src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift [hunk 5]
```diff
@@ -49,15 +118,290 @@ class CreateAntigenTestProfileViewController: UIViewController, FooterViewHandli
...
+	
+	@objc
+	private func phoneNumberTextFieldDidChange(textField: UITextField) {
+		viewModel.antigenTestProfile.phoneNumber = textField.text
+	}
+	
+	@objc
+	private func emailAddressTextFieldDidChange(textField: UITextField) {
+		viewModel.antigenTestProfile.email = textField.text
+	}
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift:227]: I think we can be much more simple if you use a KeyPath on the ViewModel, something lie ` func update(_ text: String?, keyPath: WritableKeyPath<DiaryAddAndEditEntryModel, String>)`

File: src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift [hunk 6]
```diff
@@ -49,15 +118,290 @@ class CreateAntigenTestProfileViewController: UIViewController, FooterViewHandli
...
+		dateFormatter.dateStyle = .medium
+		dateFormatter.timeStyle = .none
+		dateFormatter.locale = Locale(identifier: "de_DE")
+	}
+	
+	private func setupBindings() {
+		viewModel.$antigenTestProfile
+			.sink { [weak self] antigenTestProfile in
+				let isSaveButtonEnabled =
+						!(antigenTestProfile.firstName?.isEmpty ?? true) ||
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift:168]: Would expect the logic for enabling the button in the viewmodel. This way it also can be tested.

File: src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift [hunk 7]
```diff
@@ -49,15 +118,290 @@ class CreateAntigenTestProfileViewController: UIViewController, FooterViewHandli
...
+		viewModel.antigenTestProfile.phoneNumber = textField.text
+	}
+	
+	@objc
+	private func emailAddressTextFieldDidChange(textField: UITextField) {
+		viewModel.antigenTestProfile.email = textField.text
+	}
+}
 
+private class DescriptionCell: UITableViewCell {
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift:230]: please use extra files

File: src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift [hunk 8]
```diff
@@ -49,15 +118,290 @@ class CreateAntigenTestProfileViewController: UIViewController, FooterViewHandli
...
+			label.bottomAnchor.constraint(lessThanOrEqualTo: contentView.bottomAnchor, constant: -2)
+		])
+	}
+	
+	required init?(coder: NSCoder) {
+		fatalError("init(coder:) has not been implemented")
+	}
+}
+
+private class InputCell: UITableViewCell {
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift:260]: separate file?

File: src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift [hunk 9]
```diff
@@ -49,15 +118,290 @@ class CreateAntigenTestProfileViewController: UIViewController, FooterViewHandli
...
+		// date
+		dateFormatter = DateFormatter()
+		dateFormatter.dateStyle = .medium
+		dateFormatter.timeStyle = .none
+		dateFormatter.locale = Locale(identifier: "de_DE")
+	}
+	
+	private func setupBindings() {
+		viewModel.$antigenTestProfile
+			.sink { [weak self] antigenTestProfile in
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift:166]: why inject the `antigenTestProfile` you could just mure the logic to the viewModel and submit a bool it the input is valid or not ti enable the button

File: src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift [hunk 10]
```diff
@@ -49,15 +118,290 @@ class CreateAntigenTestProfileViewController: UIViewController, FooterViewHandli
...
+	}
+}
 
+private class DescriptionCell: UITableViewCell {
+	
+	static let identifier = "DescriptionCell"
+		
+	override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
+		super.init(style: style, reuseIdentifier: reuseIdentifier)
+		// self
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift:236]: Comments like that doesn't add any value. Please add more context to the comment. Or remove it. Please also see the other comments.

File: src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift [hunk 11]
```diff
@@ -49,15 +118,290 @@ class CreateAntigenTestProfileViewController: UIViewController, FooterViewHandli
...
+		selectionStyle = .none
+		// label
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift:238]: No value comment. Can be removed.

File: src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift [hunk 12]
```diff
@@ -49,15 +118,290 @@ class CreateAntigenTestProfileViewController: UIViewController, FooterViewHandli
...
+	}
+	
+	required init?(coder: NSCoder) {
+		fatalError("init(coder:) has not been implemented")
+	}
+}
+
+private class InputCell: UITableViewCell {
+	
+	static let identifier = "InputCell"
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift:262]: use the protocol `ReuseIdentifierProviding`?

File: src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift [hunk 13]
```diff
@@ -49,15 +118,290 @@ class CreateAntigenTestProfileViewController: UIViewController, FooterViewHandli
...
+	
+	@objc
+	private func emailAddressTextFieldDidChange(textField: UITextField) {
+		viewModel.antigenTestProfile.email = textField.text
+	}
+}
 
+private class DescriptionCell: UITableViewCell {
+	
+	static let identifier = "DescriptionCell"
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift:232]: use `ReuseIdentifierProviding`?

**Review by KaiTeuber — CHANGES_REQUESTED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift (hunk 7 above)
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift:230]: Can we have an own file for this cell?

File: src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift (hunk 13 above)
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewController.swift:232]: Marks are missing.

**Review by ArturFriesen — CHANGES_REQUESTED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/Cells/CreateAntigenTestProfileDescriptionCell.swift [hunk 14]
```diff
@@ -0,0 +1,33 @@
+////
+// 🦠 Corona-Warn-App
+//
+
+import UIKit
+
+class CreateAntigenTestProfileDescriptionCell: UITableViewCell, ReuseIdentifierProviding {
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/Cells/CreateAntigenTestProfileDescriptionCell.swift:7]: MARKs

File: src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/Cells/CreateAntigenTestProfileInputCell.swift [hunk 15]
```diff
@@ -0,0 +1,153 @@
+////
+// 🦠 Corona-Warn-App
+//
+
+import UIKit
+
+class CreateAntigenTestProfileInputCell: UITableViewCell, ReuseIdentifierProviding {
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/Cells/CreateAntigenTestProfileInputCell.swift:7]: MARKs

**Review by KaiTeuber — APPROVED** [review]

**Review by KaiTeuber — CHANGES_REQUESTED** [review]
Please add Unittest for the view models

**Review by ArturFriesen — APPROVED** [review]

**Review by KaiTeuber — APPROVED** [review]

**Review by janetback — APPROVED** [review]
UA reviewed

**Pull request merged**